File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-3.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        l = r = 0
        res = 1
        chars = defaultdict(int)
        fr = s[l]
        while r < len(s):
            chars[s[r]] += 1
            if chars[s[r]] > chars[fr]:
                fr = s[r]
            while r-l+1 - chars[fr] > k:
                chars[s[l]] -= 1
                l += 1
                for i in range(l, r+1):
                    if chars[s[i]] > chars[fr]:
                        fr = s[i]
            else:
                res = max(res, r-l+1)
                r += 1
        return res

File: Data Structures & Algorithms/longest-turbulent-subarray/submission-3.rs
impl Solution {
    pub fn max_turbulence_size(arr: Vec<i32>) -> i32 {
        let mut res = 1;
        let mut last: Option<bool> = None;
        let mut l = 0;
        for r in 1..arr.len() {
            let next = if arr[r-1] > arr[r] { Some(true) } else if arr[r-1] < arr[r] { Some(false) } else { None };
            if next.is_none() || last == next {
                l = if next.is_none() {r} else {r-1};
            }

            last = next;
            res = res.max((r-l+1) as i32);
        }

        res
    }
}
